### ISSUE TITLE
completion and predictive tags

### ISSUE THREAD

**Issue by ztmr** [olahol/react-tagsinput#4]
Do you plan any completion/predictions? I am thinking whether to hack your component or write my own. I think the easiest way of enabling your component for completion is to add something like `onBeforeTagAdd` and `onBeforeTagRemove` (and probably rename `onTag{Add,Remove}` to `onAfterTag{Add,Remove}`) so that the completion can be made on demand before the tag is added.
`onBeforeTagRemove` might be worth for the case where somebody does not want to allow removing tag X that is a prerequisite for another tag Y or so.

The use case in my head is "multiselect on steroids." Imagine the form like this:
- one tag-input for user group -- choose tags where tag = user group id (name is displayed, id is internal)
- one tag-input for users -- choose users (tag = user, name displayed, id is internal), but only from the members of selected groups

...in other words, tag should be an object, not a string, so there should be also some method for "matching" the objects agains the input keyword :)

It would be also nice to have `onAfterAddKeys` so that enter on empty tag would trigger some action (for example skip to the next input in the form).

Is that something where your component is going to evolve in the future? ;-)

**Comment by olahol** [comment]
This is an interesting proposal. I did not have any plans to implement completion but now that you mention it I think it would be a good idea to at least add the methods `onBeforeTag{Add, Remove}` and `onAfterTag{Add, Remove}` so that people can implement it themselves. 

I do still think that the internal representation of tags should be strings though as it represent what is actually shown in the component and that more complex object could be handled by a parent component. I also want to keep `onTag{Add, Remove}` for backwards compatibility reasons.

**Comment by ztmr** [comment]
I don't think it's necessary to have full objects stored in the internal state. I would simply introduce a function (property) that does mapping of the internal string to something to be displayed to the user. So internally, you could have `"123"` or `"asdf"` and the tag "pill" might show up like `"[123] John Doe"` or `"[asdf]"`. That function should be `x => x` by default, of course.

If somebody wanted object representation, then it's no problem to do `TagInput.getTags().map(tagId => myTagObjects[tagId])` or so...

The `onAfterAddKeys` to handle "confirm" in the sense "done, user does not want to add anything, let's go for the next form input" would be great to have too :)

I am thinking about completion now... if we have the `onBeforeTagAdd`, then I would be able to display some hints under the TagsInput, but if the user clicks the hint we will need to pass the chosen tag back to the TagsInput. So there should be some callback in argument or at least `TagsInput.{add,remove}Tag` to update TagsInput externally. Or the wrapping component should force TagsInput to re-mount or at least to receive new props with the original tags modified. Hm? Just loudly thinking... What do you think?

**Comment by olahol** [comment]
I have begun a draft of completion support on the branch [completion](https://github.com/olahol/react-tagsinput/tree/completion). 

I have added the `onBeforeTag` props which return true/false on whether a tag should be added/removed. I have also modified the internal methods `addTag` and `removeTag` to take a tag as input so tags can be dynamically added/removed.

There is more work to be done and if you have any ideas I am more than happy to merge pull requests :smile:
